components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/ConfigurationFacade.java

```diff
@@ -355,13 +358,22 @@ private String buildAbsoluteUrl(String defaultContext, Supplier<String> getValue
                 path = urlFromFileBasedConfig;
             }
         }
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        String organizationId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getOrganizationId();
+        ServiceURLBuilder serviceURLBuilder = ServiceURLBuilder.create().addPath(path);
         try {
-            String organizationId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getOrganizationId();
-            return ServiceURLBuilder.create().addPath(path).setOrganization(organizationId).build()
-                    .getAbsolutePublicURL();
+            // If the organization ID is not set in the context for a tenant which is associated to an organization.
+            if (StringUtils.isEmpty(organizationId) && OrganizationManagementUtil.isOrganization(tenantDomain)) {
```

Comment:
This service URL builder logics are only executed when resolving the URLs for password recovery and self registration flows. The authentication endpoint related service URLs build from different place, hence there should not be any impact for the login flow due to this introduced check.
Validated with console app is deployed in IS and static app from outside also whether the authentication flow get executed these introduced logics. Only recovery flows will get executed this check.